src/agent/discovery_proxy.cpp

```diff
@@ -0,0 +1,369 @@
+    instanceInfo.mTxtData   = aInstanceInfo.mTxtData.data();
+    instanceInfo.mTtl       = std::min(aInstanceInfo.mTtl, static_cast<uint32_t>(kServiceTtlCapLimit));
+
+    std::for_each(mSubscriptions.begin(), mSubscriptions.end(), [&](const MdnsSubscription &aSubscription) {
+        if (aSubscription.MatchesServiceInstance(aType, aInstanceInfo.mName))
+        {
+            std::string serviceFullName  = aType + "." + aSubscription.mDomain;
+            std::string hostName         = TranslateDomain(aInstanceInfo.mHostName, aSubscription.mDomain);
+            std::string instanceFullName = aInstanceInfo.mName + "." + serviceFullName;
+
+            instanceInfo.mFullName = instanceFullName.c_str();
+            instanceInfo.mHostName = hostName.c_str();
+
+            otDnssdQueryNotifyServiceInstance(mNcp.GetInstance(), serviceFullName.c_str(), &instanceInfo);
+        }
+    });
+}
+
+void DiscoveryProxy::OnHostDiscovered(const std::string &                        aHostName,
+                                      const Mdns::Publisher::DiscoveredHostInfo &aHostInfo)
+{
+    otDnssdHostInfo hostInfo;
+
+    otbrLog(OTBR_LOG_INFO, "[dnssd discovery proxy] host discovered: %s hostname %s address %s", aHostName.c_str(),
+            aHostInfo.mHostName.c_str(), aHostInfo.mAddress.ToString().c_str());
+
+    CheckHostnameSanity(aHostInfo.mHostName);
+
+    hostInfo.mAddress = *reinterpret_cast<const otIp6Address *>(&aHostInfo.mAddress);
+    hostInfo.mTtl     = std::min(aHostInfo.mTtl, static_cast<uint32_t>(kServiceTtlCapLimit));
+
+    std::for_each(mSubscriptions.begin(), mSubscriptions.end(), [&](const MdnsSubscription &aSubscription) {
+        if (aSubscription.MatchesHost(aHostName))
+        {
+            std::string hostFullName = TranslateDomain(aHostInfo.mHostName, aSubscription.mDomain);
+
+            otDnssdQueryNotifyHost(mNcp.GetInstance(), hostFullName.c_str(), &hostInfo);
+        }
+    });
+}
+
+std::string DiscoveryProxy::TranslateDomain(const std::string &aName, const std::string &aTargetDomain)
+{
+    std::string targetName;
+    std::string hostName, domain;
+
+    VerifyOrExit(OTBR_ERROR_NONE == SplitFullHostName(aName, hostName, domain), targetName = aName);
+    VerifyOrExit(domain == "local.", targetName = aName);
+
+    targetName = hostName + "." + aTargetDomain;
+
+exit:
+    otbrLog(OTBR_LOG_DEBUG, "Translate domain: %s => %s", aName.c_str(), targetName.c_str());
+    return targetName;
+}
+
+int DiscoveryProxy::GetServiceSubscriptionCount(const std::string &aInstanceName, const std::string &aServiceName)
+{
+    return std::accumulate(
+        mSubscriptions.begin(), mSubscriptions.end(), 0, [&](int aAccum, const MdnsSubscription &aSubscription) {
+            return aAccum + (aSubscription.mInstanceName == aInstanceName && aSubscription.mServiceName == aServiceName
+                                 ? aSubscription.mSubscriptionCount
+                                 : 0);
```

Comment:
I think there should only one subscription that has the same instance and service name, right? Seems we don't to accumulate.

Reply:
There can be duplicate subscription from OT core to Discovery Proxy.